lexers/nix.lua

```diff
@@ -0,0 +1,81 @@
+-- Nix LPEG lexer.
+-- Written by Samuel Marquis.
+
+local lexer = lexer
+local P, S = lpeg.P, lpeg.S
+
+local lex = lexer.new(...)
+
+-- Keywords.
+lex:add_rule('keyword', lex:tag(lexer.KEYWORD, lex:word_match(lexer.KEYWORD)))
+
+-- Functions.
+lex:add_rule('function', lex:tag(lexer.FUNCTION_BUILTIN, lex:word_match(lexer.FUNCTION_BUILTIN)))
+
+-- Constants.
+lex:add_rule('constant', lex:tag(lexer.CONSTANT_BUILTIN, lex:word_match(lexer.CONSTANT_BUILTIN)))
+
+-- Strings.
+local str = lexer.range('"', true)
+local ml_str = lexer.range("''", false)
+lex:add_rule('string', lex:tag(lexer.STRING, str + ml_str))
+
+-- Paths.
+local path_char = lexer.alnum + S('_-.+')
+local path_seg = ('/' * path_char^1)
+local path = P('~')^-1 * path_char^0 * path_seg^1 * lpeg.P('/')^-1
+lex:add_rule('path', lex:tag(lexer.LINK, path))
+
+-- URIs.
+local uri_char = lexer.alnum + S("%/?:@&=+$,-_.!~*'")
+local uri = lexer.alpha * (lexer.alnum + S('+-.'))^0 * ':' * uri_char^1
+lex:add_rule('uri', lex:tag(lexer.LINK, uri))
+
+-- Angle-bracket paths.
+local spath = '<' * path_char^1 * path_seg^0 * '>'
+lex:add_rule('spath', lex:tag(lexer.LINK, spath))
+
+-- Identifiers.
+local id = (lexer.alpha + '_') * (lexer.alnum + S('_-'))^0
+lex:add_rule('identifier', lex:tag(lexer.IDENTIFIER, id))
+
+-- Comments.
+local line_comment = lexer.to_eol('#', true)
+local block_comment = lexer.range('/*', '*/')
+lex:add_rule('comment', lex:tag(lexer.COMMENT, line_comment + block_comment))
+
+-- Numbers.
+lex:add_rule('number', lex:tag(lexer.NUMBER, lexer.number))
+
+-- Operators.
+local ops = S('?+-.*/!<>=,;:()[]{}') + lex:word_match(lexer.OPERATOR)
```

Comment:
Please use `P('&&') + '||' + ...` instead of word match because these characters will be added to an internal word character list and result in more cases of potential word matches and then backtracking on failure.

Also, if you really want to match '//', '->', and '++' on their own, they needs to be matched before the `S()` or else '//' will be lexed as two separate '/' characters.

Reply:
So does changing the order of addition here provide this effect, or do I want separate `add_rule()` calls for the long/short operators?

Reply:
Yes, switching the operands to `+` is sufficient. It's an ordered choice operator, so it tries to match left first. If that fails, it tries to match right.

Reply:
Is it left-associative? Wrt `P() + "..." + S()`

Reply:
Yes, I think so.

Reply:
Solid, refactored this